### DISCUSSION TITLE
Simple Time Sheet

### DISCUSSION THREAD

**Discussion by beccon4** [freeplane/freeplane#1091] [Help] [answered]
To include simple time management and controlling into a mindmap I'd like to calculate and later aggregate the time spent on a particular item. 

To do so, I add attribute for the begin and end time. This need not to be a complete date as the day doesn't matter - only the hours. For that I had to serialise them in a simple way - such as minutes after midnight.

Is there a more elegant way of parsing 24h time values than using substring? (or regexp)

BTW - not even this works for some strange reason:

`= (24*node["end"].substring(0,2).to.num+node["end"].substring(3,2).to.num) - (24*node["begin"].substring(0,2).to.num+node["begin"].substring(3,2).to.num)
`
I won't be able to do a simple conversion of the substring to numeric to be able to apply calculations on:

`= node["end"].substring(0,2).to.num`

while

`= node["end"].substring(0,2)`

works as expected but results in a string.

What's wrong?

**Comment by sifran-github** [comment, marked as answer]
This is working:
`=(60*node["end"].substring(0,2).toInteger()+node["end"].substring(3,5).toInteger()) - (60*node["begin"].substring(0,2).toInteger()+node["begin"].substring(3,5).toInteger())`

notice:
1. it is 60, not 24.
2. .to.num -> toInteger() worked for me.
3. .substring(x,y): y is the end index, NOT length.
4. i couldnt make an attribute "20:30" just as text, without freeplane converting it into a full date. its a bit strange. so instead i used "20-30" or just any string which is not getting converted automatically. otherwise if the format is HH:mm it will still return "20" because the full value starts with "2023..." even you dont see it.

**Reply by beccon4** [reply]
1. "it is 60, not 24." :facepalm: 

2. "to.num -> toInteger() worked for me."
For me too - thank you very much.

4. There is a global option In Settings-> Behaviour -> format and recognize date values (do not know the exact terms) where one can change the behavior accordingly. I'd prefer to prefix the time value with ' as I know it from Excel. All characters would shift one position to the right though.

**Reply by bepolymathe** [reply]
This is a great idea and I would be very interested if you could share the results of your research.

**Comment by EdoFro** [comment]
> Is there a more elegant way of parsing 24h time values than using substring? (or regexp)

using the same logic but not using substring:
```groovy
=
def beginDate = node['begin'].to.date
def endDate   = node['end'  ].to.date

return (endDate.hours*60 + endDate.minutes) - (beginDate.hours*60 + beginDate.minutes)
```

even "fancier" would be to use the [TimeCategory](https://docs.groovy-lang.org/latest/html/api/groovy/time/TimeCategory.html):
```groovy
=
def beginDate = node['begin'].to.date
def endDate   = node['end'  ].to.date

def duration
use(groovy.time.TimeCategory) {
    duration = endDate - beginDate
}

return duration
```

where duration is a [TimeDuration object](https://docs.groovy-lang.org/latest/html/api/groovy/time/TimeDuration.html)

BR

edo
